zcash_primitives/src/transaction/builder.rs

```diff
@@ -625,55 +961,36 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
     /// Builds a transaction from the configured spends and outputs.
     ///
     /// Upon success, returns a tuple containing the final transaction, and the
-    /// [`TransactionMetadata`] generated during the build process.
+    /// [`SaplingMetadata`] generated during the build process.
     ///
     /// `consensus_branch_id` must be valid for the block height that this transaction is
     /// targeting. An invalid `consensus_branch_id` will *not* result in an error from
     /// this function, and instead will generate a transaction that will be rejected by
     /// the network.
     pub fn build(
         mut self,
+        version: TxVersion,
```

Comment:
Instead of passing in the transaction version, the builder should use `target_height` to decide what version of transaction to build. Then this can be stored in the builder, and we can return errors if e.g. the user tries to call `add_orchard_output` for a transaction targeting pre-NU5.

If we eventually add Sprout support to the Rust transaction builder (or in future deprecate some other pool) then we will need the ability to choose the kind of transaction, but that doesn't necessarily mean specifying the precise transaction version used (I think the builder should just pick the best version for the caller's expressed requirements).